File: 课程网站/ch3/code/python/3.1-fourier-forms.py
"""3.1 四种傅里叶变换：同一类脉冲的四种表示
课程站 ch3 页内 Python 片段的可运行文件版。2026-09-18 本机验证通过。
运行：python 本文件；需要 numpy、matplotlib（3.1）、scipy（3.7）。
"""
import numpy as np, matplotlib.pyplot as plt
width, T, N = 3, 8, 8
t = np.linspace(-16, 16, 6401); Om = np.linspace(-4*np.pi, 4*np.pi, 3201)
w = Om.copy(); n = np.arange(-16, 17)
m = np.arange(-int(4*np.pi/(2*np.pi/T)), int(4*np.pi/(2*np.pi/T))+1); k = np.arange(-2*N, 2*N+1)
sinc_u = lambda z: np.sinc(z/np.pi)            # np.sinc 是 sin(pi z)/(pi z)，换算成 sin(z)/z
p = (np.abs(t) < width/2).astype(float)
P = width*sinc_u(Om*width/2)
c = (width/T)*sinc_u(m*(2*np.pi/T)*width/2)
x = (np.abs(n) <= 1).astype(float); X = 1+2*np.cos(w)
Xdfs = 1+2*np.cos(2*np.pi*k/N)
print("P(0)=%.3f  c0=%.4f  X(pi)=%.3f" % (P[np.argmin(abs(Om))], c[m == 0][0], 1+2*np.cos(np.pi)))
fig, ax = plt.subplots(4, 2, figsize=(12, 10))
ax[0,0].plot(t, p); ax[0,1].plot(Om/np.pi, np.abs(P))
ax[1,0].plot(t, (np.abs(np.mod(t+T/2, T)-T/2) < width/2).astype(float)); ax[1,1].stem(m*(2*np.pi/T)/np.pi, np.abs(c))
ax[2,0].stem(n, x); ax[2,1].plot(w/np.pi, np.abs(X))
ax[3,0].stem(n, np.isin(np.mod(n, N), [0, 1, N-1]).astype(float)); ax[3,1].stem(2*k/N, np.abs(Xdfs))
for a, ttl in zip(ax.flat, ["CTFT: p(t)", "|P(jΩ)|", "FS: p_T(t), T=8", "|c_m|", "DTFT: x[n]", "|X(e^jω)|", "DFS: N=8", "|X_DFS[k]|"]):
    a.set_title(ttl); a.grid(True)
plt.tight_layout(); plt.show()

File: 课程网站/ch3/code/julia/3.1-fourier-forms.jl
# 3.1 四种傅里叶变换：连续周期脉冲的 FS 系数、三点序列的 DTFT、周期序列的 DFS
# 课程站 ch3 的 MWORKS Syslab（Julia）对照版，与 demo_fourier_forms.m / 3.1-fourier-forms.py 同一算法。只用 Julia 标准库。
T = 8.0; width = 3.0                       # 连续周期矩形脉冲：周期 T，宽度 3
c0 = width / T                              # 直流系数 c0 = 宽度/T
ck(k) = k == 0 ? c0 : sin(pi * k * width / T) / (pi * k)   # FS 系数（幅度 1 的脉冲）
println("FS: c0 = ", round(c0, digits = 4), "  c1 = ", round(ck(1), digits = 4), "  c2 = ", round(ck(2), digits = 4))
x = Dict(-1 => 1.0, 0 => 1.0, 1 => 1.0)     # 三点序列 x[-1]=x[0]=x[1]=1
X(w) = sum(v * exp(-im * w * n) for (n, v) in x)          # DTFT：对连续 ω 定义，2π 周期
println("DTFT: X(0) = ", round(real(X(0.0)), digits = 4), "  X(π/2) = ", round(real(X(pi / 2)), digits = 4), "  X(π) = ", round(real(X(pi)), digits = 4), "  X(2π) = ", round(real(X(2pi)), digits = 4))
N = 8; xp = [1.0, 1.0, 0, 0, 0, 0, 0, 1.0]  # 一个周期：索引 0、1、7 处为 1
Xdfs(k) = sum(xp[n + 1] * exp(-2im * pi * k * n / N) for n in 0:N-1)   # DFS
vals = [round(real(Xdfs(k)), digits = 4) for k in (0, 2, 4)]
println("DFS: X~(0), X~(2), X~(4) = ", vals)
@assert abs(c0 - 0.375) < 1e-12 && abs(real(X(0.0)) - 3) < 1e-12 && abs(real(X(pi)) + 1) < 1e-12
@assert vals == [3.0, 1.0, -1.0] && abs(X(2pi) - X(0.0)) < 1e-12
println("FOURIER_FORMS_OK")
